Adopt Safer CPP in ImageAnalysisQueue.cpp

jelee53 opened the pull request:
#### 62edb8ae1af79efecc32759d192b1f5795afe4a6
<pre>
Adopt Safer CPP in ImageAnalysisQueue.cpp
<a href="https://bugs.webkit.org/show_bug.cgi?id=304617">https://bugs.webkit.org/show_bug.cgi?id=304617</a>
<a href="https://rdar.apple.com/167049077">rdar://167049077</a>

Reviewed by Ryosuke Niwa.

Apply <a href="https://github.com/WebKit/WebKit/wiki/Safer-CPP-Guidelines.">https://github.com/WebKit/WebKit/wiki/Safer-CPP-Guidelines.</a>

No new tests needed.

* Source/WebCore/SaferCPPExpectations/UncheckedLocalVarsCheckerExpectations:
* Source/WebCore/SaferCPPExpectations/UncountedCallArgsCheckerExpectations:
* Source/WebCore/SaferCPPExpectations/UncountedLambdaCapturesCheckerExpectations:
* Source/WebCore/SaferCPPExpectations/UncountedLocalVarsCheckerExpectations:
* Source/WebCore/page/ImageAnalysisQueue.cpp:
(WebCore::ImageAnalysisQueue::enqueueAllImagesRecursive):
(WebCore::ImageAnalysisQueue::resumeProcessing):

Canonical link: <a href="https://commits.webkit.org/305397@main">https://commits.webkit.org/305397@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/30ba0a5be8afaaab019acc9aaed2a05a56bb3cc7

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/138277 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/10642 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/49687 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/146357 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/91252 "Built successfully") | ⏳ 🛠 ios-apple 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/140151 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/11346 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/10796 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/105784 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/77158 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/a841a6ad-7350-4fb1-bac1-b5ffa34a69f7) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/141224 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/162/builds/8500 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/123962 "Passed tests") | [❌ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/86630 "Found 1 new API test failure: WPE/TestWebKitWebXR:/webkit/WebKitWebXR/leave-immersive-mode (failure)") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/25ccb891-2204-44fd-b1bc-115930c37cb4) 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/154/builds/8090 "Passed tests") | [✅ 🧪 api-mac-debug](https://ews-build.webkit.org/#/builders/165/builds/5853 "Passed tests") | [✅ 🛠 wpe-cairo-libwebrtc](https://ews-build.webkit.org/#/builders/166/builds/6638 "Built successfully") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/117503 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/42162 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/149068 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/10324 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/42719 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/114187 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/10341 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/8724 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/114528 "Passed tests") | | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/8062 "Passed tests") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/120249 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/65147 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/21296 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/10371 "Built successfully") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/38183 "Passed tests") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/10102 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/73938 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/10311 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/10162 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

annevk commented on Source/WebCore/page/ImageAnalysisQueue.cpp:
```diff
@@ -163,14 +163,15 @@ void ImageAnalysisQueue::resumeProcessing()
 
         m_pendingRequestCount++;
-        m_page->resetTextRecognitionResult(*element);
+        RefPtr webPage = m_page.get();
```
```suggestion
        Ref webPage = *m_page;
```

annevk commented on Source/WebCore/page/ImageAnalysisQueue.cpp:
```diff
@@ -163,14 +163,15 @@ void ImageAnalysisQueue::resumeProcessing()
         auto allowSnapshots = m_languageIdentifiers.target.isEmpty() ? TextRecognitionOptions::AllowSnapshots::Yes : TextRecognitionOptions::AllowSnapshots::No;
-        m_page->chrome().client().requestTextRecognition(*element, { m_languageIdentifiers.source, m_languageIdentifiers.target, allowSnapshots }, [this, page = m_page] (auto&&) {
-            if (!page || page->imageAnalysisQueueIfExists() != this)
+        webPage->chrome().client().requestTextRecognition(*element, { m_languageIdentifiers.source, m_languageIdentifiers.target, allowSnapshots }, [this, protectedThis = Ref { *this }, webPage] (auto&&) {
```
```suggestion
        webPage->chrome().client().requestTextRecognition(*element, { m_languageIdentifiers.source, m_languageIdentifiers.target, allowSnapshots }, [this, protectedThis = Ref { *this }, webPage](auto&&) {
```

annevk commented on Source/WebCore/page/ImageAnalysisQueue.cpp:
```diff
@@ -163,14 +163,15 @@ void ImageAnalysisQueue::resumeProcessing()
 
         m_pendingRequestCount++;
-        m_page->resetTextRecognitionResult(*element);
+        Ref webPage = *m_page;
```
I would call this `page`.

annevk commented on Source/WebCore/page/ImageAnalysisQueue.cpp:
```diff
@@ -163,14 +163,15 @@ void ImageAnalysisQueue::resumeProcessing()
-        m_page->chrome().client().requestTextRecognition(*element, { m_languageIdentifiers.source, m_languageIdentifiers.target, allowSnapshots }, [this, page = m_page] (auto&&) {
-            if (!page || page->imageAnalysisQueueIfExists() != this)
+        webPage->chrome().client().requestTextRecognition(*element, { m_languageIdentifiers.source, m_languageIdentifiers.target, allowSnapshots }, [this, protectedThis = Ref { *this }, webPage](auto&&) {
+            if (webPage->imageAnalysisQueueIfExists() != this)
```
I think the result here is that now we keep page alive, I'm not sure that's what we want.

Probably best to ask @cdumez or maybe Youenn or rniwa.

cdumez replied:
Indeed, this is a behavior change and it is safer to maintain the pre-existing behavior. We should capture a WeakPtr to the page and convert it to a RefPtr inside the lambda.

rniwa commented on Source/WebCore/page/ImageAnalysisQueue.cpp:
```diff
@@ -163,14 +163,16 @@ void ImageAnalysisQueue::resumeProcessing()
-        m_page->chrome().client().requestTextRecognition(*element, { m_languageIdentifiers.source, m_languageIdentifiers.target, allowSnapshots }, [this, page = m_page] (auto&&) {
-            if (!page || page->imageAnalysisQueueIfExists() != this)
+        page->chrome().client().requestTextRecognition(*element, { m_languageIdentifiers.source, m_languageIdentifiers.target, allowSnapshots }, [this, protectedThis = Ref { *this }, weakPage = WeakPtr { page }](auto&&) {
+            RefPtr webPage = weakPage.get();
```
It's very confusing to call this variable "webPage" since there is also `WebPage` class defined in `WebKitLegacy` and `WebKit`. Maybe call it `protectedPage`?

annevk replied:
I would have gone with `page` again. Using `protected` in a local variable is quite strange. I thought that was only for `protectedThis`.

jelee53 committed:
Adopt Safer CPP in ImageAnalysisQueue.cpp
https://bugs.webkit.org/show_bug.cgi?id=304617
rdar://167049077

Reviewed by Ryosuke Niwa.

Apply https://github.com/WebKit/WebKit/wiki/Safer-CPP-Guidelines.

No new tests needed.

* Source/WebCore/SaferCPPExpectations/UncheckedLocalVarsCheckerExpectations:
* Source/WebCore/SaferCPPExpectations/UncountedCallArgsCheckerExpectations:
* Source/WebCore/SaferCPPExpectations/UncountedLambdaCapturesCheckerExpectations:
* Source/WebCore/SaferCPPExpectations/UncountedLocalVarsCheckerExpectations:
* Source/WebCore/page/ImageAnalysisQueue.cpp:
(WebCore::ImageAnalysisQueue::enqueueAllImagesRecursive):
(WebCore::ImageAnalysisQueue::resumeProcessing):

Canonical link: https://commits.webkit.org/305397@main

cdumez commented on Source/WebCore/page/ImageAnalysisQueue.cpp:
```diff
@@ -163,14 +163,16 @@ void ImageAnalysisQueue::resumeProcessing()
-        m_page->chrome().client().requestTextRecognition(*element, { m_languageIdentifiers.source, m_languageIdentifiers.target, allowSnapshots }, [this, page = m_page] (auto&&) {
-            if (!page || page->imageAnalysisQueueIfExists() != this)
+        page->chrome().client().requestTextRecognition(*element, { m_languageIdentifiers.source, m_languageIdentifiers.target, allowSnapshots }, [this, protectedThis = Ref { *this }, weakPage = WeakPtr { page }](auto&&) {
+            RefPtr protectedPage = weakPage.get();
```
Yes, ideally this would have been `page`.
